src/hip_memory.cpp

```diff
@@ -320,20 +320,31 @@ void generic_copy(void* __restrict dst, const void* __restrict src, size_t n,
 inline
 void memcpy_impl(void* __restrict dst, const void* __restrict src, size_t n,
                  hipMemcpyKind k) {
+    auto si{info(src)};
+    auto di{info(dst)};
+
+    if (!is_large_BAR){
```

Comment:
Maybe add a comment as to why we are doing this.. something like direction determined by src/dst pointer info takes precedence over hipMemcpyKind if there is a mismatch in input parameters passed to the API.
Also why is this applicable only for small BAR systems?

Reply:
It does not fail for large BAR hence the check. Mostly failure happens because there are cases where lower functions uses std::memcpy without checking if pointers are host pointer.